### PR TITLE
v3.4.0-RC.3

### PR DESCRIPTION
**Pull request by cgcardona**
## Changes

[Perform Apricot Phase 2 Validation on EVM ImportTxs](https://github.com/ava-labs/avalanchejs/pull/187)

## Details

**TLDR**—in EVM `ImportTx` we’re now enforcing:

1. Each `EVMOutput` is unique for an `address` and `assetID`. This means that even if the `ImportTx` has 2 AVAX inputs and 2 ANT inputs it can only have 1 `EVMOutput` for each `assetID` per `address`
2. The importing transaction must burn enough AVAX to cover a transaction fee

### Validating Uniqueness

First, we're [creating a `Map`](https://github.com/ava-labs/avalanchejs/blob/development/src/apis/evm/importtx.ts#L253). Then we're looping over the outputs and mapping addresses to assetIDs. If there are multiple outputs with the same address and assetID then we [throw an error](https://github.com/ava-labs/avalanchejs/blob/development/src/apis/evm/importtx.ts#L261).

```ts
// Example Map
Map(2) {
  '0x8db97C7cEcE249c2b98bDC0226Cc4C2A57BF52FC' => [
    'FvwEAhmxKfeiG8SnEvq42hc6whRyY3EFYAvebMqDNDGCgxN5Z',
    'F4MyJcUvq3Rxbqgd4Zs8sUpvwLHApyrp4yxJXe2bAV86Vvp38'
  ],
  '0xecC3B2968B277b837a81A7181e0b94EB1Ca54EdE' => [
    'FvwEAhmxKfeiG8SnEvq42hc6whRyY3EFYAvebMqDNDGCgxN5Z',
    '2Df96yHyhNc3vooieNNhyKwrjEfTsV2ReMo5FKjMpr8vwN4Jqy',
    'SfSXBzDb9GZ9R2uH61qZKe8nxQHW9KERW9Kq9WRe4vHJZRN3e'
  ]
}
```

### Validating Fee

For the fee we're [summing all of the inputs](https://github.com/ava-labs/avalanchejs/blob/development/src/apis/evm/importtx.ts#L273) and then we're [subtracting all of the outputs](https://github.com/ava-labs/avalanchejs/blob/development/src/apis/evm/importtx.ts#L282) which gives us the fee. If the fee is less than the required fee we [throw an error](https://github.com/ava-labs/avalanchejs/blob/development/src/apis/evm/importtx.ts#L290).

### Automated

* Single AVAX EVMOutput fail
* Single AVAX EVMOutput success
* Multiple AVAX EVMOutput fail
* Multiple AVAX EVMOutput success
* Single ANT EVMOutput fail
* Single ANT EVMOutput success
* Multiple ANT EVMOutput fail
* Multiple ANT EVMOutput success

### Manual

* [evm-single-avax-import-fail.ts](https://gist.github.com/cgcardona/d3c8390ded1d38d8abe457a0b19b71dc)
* [evm-single-avax-import-success.ts](https://gist.github.com/cgcardona/b082daaf7a81ee2efa4669890517c365)
* [evm-multi-avax-import-fail.ts](https://gist.github.com/cgcardona/0427e1b66e0b2cd382db7561dd671843)
* [evm-multi-avax-import-success.ts](https://gist.github.com/cgcardona/a0cb63e0f8c5d6769014b18e4664a5a9)
* [evm-multi-asset-import-fail.ts](https://gist.github.com/cgcardona/849e7db4471136d9e8fb6ba67b16ea70)
* [evm-multi-asset-import-success.ts](https://gist.github.com/cgcardona/c6a04eeec5d8e9f2ad5d7b47de0521d7)

## Miscellaneous

* Cleaning up the example scripts' syntax
  * Remove commas
  * Remove unneeded `await`s
  * Break multi-component `import` statements across multiple lines

## buildImportTx

In the EVM's `buildImportTx` [create a map of assetIDs to amounts](https://github.com/ava-labs/avalanchejs/blob/078bbd58be2fec4b16f26162358578b512005694/src/apis/evm/utxos.ts#L295). After the fees are paid [for each duplicate assetID increment the existing amount](https://github.com/ava-labs/avalanchejs/blob/078bbd58be2fec4b16f26162358578b512005694/src/apis/evm/utxos.ts#L338-L348). Lastly loop over the map and [create singular EVMOutputs for each assetID](https://github.com/ava-labs/avalanchejs/blob/078bbd58be2fec4b16f26162358578b512005694/src/apis/evm/utxos.ts#L338-L359).

### CODE REVIEW

**Review by EvanJRichard — APPROVED** [review]
Looking good to me!

File: examples/avm/exportTx-ant-cchain.ts [hunk 1]
```diff
@@ -57,8 +57,8 @@ const main = async (): Promise<any> => {
     let assetID: Buffer = utxo.getAssetID()
     const outputidx: Buffer = utxo.getOutputIdx()
     let secpTransferOutput: SECPTransferOutput = new SECPTransferOutput()
-    if(avaxAssetID.toString("hex") === assetID.toString("hex")) {
-      assetID = avaxAssetID
+    if(avaxAssetIDBuf.toString("hex") === assetID.toString("hex")) {
+      assetID = avaxAssetIDBuf
```
Comment (danlaine) [inline examples/avm/exportTx-ant-cchain.ts:61]: I might be missing something since I don't know typescript very well, but is this a no-op? We're checking if `avaxAssetIDBuf` is the same as `assetID`, and if they are, we set `assetID` to `avaxAssetIDBuf`, but aren't they already the same?

Comment (cgcardona) [reply examples/avm/exportTx-ant-cchain.ts]: Good catch. Resolved.

File: examples/avm/exportTx-ant-cchain.ts [hunk 2]
```diff
@@ -32,8 +32,9 @@ const privKey: string = "PrivateKey-ewoqjP7PxY4yr3iLTpLisriqt94hdyDFNgchSxGGztUr
 xKeychain.importKey(privKey)
 const xAddresses: Buffer[] = xchain.keyChain().getAddresses()
 const xAddressStrings: string[] = xchain.keyChain().getAddressStrings()
-const blockchainid: string = Defaults.network['12345'].X.blockchainID
+const blockchainID: string = Defaults.network['12345'].X.blockchainID
 const cChainBlockchainID: string = Defaults.network['12345'].C.blockchainID
+const avaxAssetIDBuf: Buffer = bintools.cb58Decode(Defaults.network['12345'].X.avaxAssetID)
```
[the remark this answers is not in the document]
Comment (cgcardona) [reply examples/avm/exportTx-ant-cchain.ts]: Resolved

File: src/apis/evm/importtx.ts [hunk 3]
```diff
@@ -222,7 +227,67 @@ export class ImportTx extends EVMBaseTx {
       if(outs.length > 1) {
         outs = outs.sort(EVMOutput.comparator());
       }
+      outputsPassed = true;
       this.outs = outs;
     }
+    if(inputsPassed && outputsPassed) {
```
[the remark this answers is not in the document]
Comment (cgcardona) [reply src/apis/evm/importtx.ts:233]: Resolved
